File: PQC/BouncyCastle/MyBcPQC/TestMyBcPQC/TestBcPqcSignatur/Stateless/Test-ML-DSA/BobMLDSA.cs
using Org.BouncyCastle.Crypto.Generators;
using Org.BouncyCastle.Crypto.Parameters;
using Org.BouncyCastle.Security;

namespace michele.natale.TestBcPqcs;

using michele.natale.BcPqcs;
using Org.BouncyCastle.Crypto.Signers;
using Services;

public class BobMLDSA : IDisposable
{
  public byte[] PubKey { get; private set; } = null!;
  public bool IsDisposed { get; private set; } = true;
  public MlDsaKeyPairInfo Info { get; private set; } = null!;
  public MLDsaParameters Parameter { get; private set; } = null!;

  private SecureRandom Rand = new();


  public BobMLDSA()
    : this(ToRngParameter())
  {
  }

  public BobMLDSA(MLDsaParameters parameter)
  {
    this.Parameter = parameter;

    var generator = new MLDsaKeyGenerationParameters(
      this.Rand, this.Parameter);

    var keypair_generator = new MLDsaKeyPairGenerator();
    keypair_generator.Init(generator);

    var keypair = keypair_generator.GenerateKeyPair();
    this.PubKey = ((MLDsaPublicKeyParameters)keypair.Public).GetEncoded();
    var privkey = ((MLDsaPrivateKeyParameters)keypair.Private).GetEncoded();
    this.Info = new MlDsaKeyPairInfo(this.PubKey, privkey, parameter);

    Array.Clear(privkey);
    this.IsDisposed = false;
  }

  public void Clear()
  {
    if (this.IsDisposed) return;

    if (this.PubKey is not null)
      Array.Clear(this.PubKey);

    this.Rand = null!;
    this.Info = null!;
    this.PubKey = null!;
    this.Parameter = null!;

  }

  public MLDsaPublicKeyParameters ToPubKey() =>
    MLDsaPublicKeyParameters.FromEncoding(
      this.Parameter, this.PubKey);


  public (byte[] Sign, byte[] PubKey) Sign(ReadOnlySpan<byte> message)
  {
    var msg = message.ToArray();

    //ML-DSA Parameter 
    var parameter = this.Info.ToParameter();

    //Sign Message-Data
    var signer = new MLDsaSigner(parameter, true);
    var privkey = MLDsaPrivateKeyParameters
      .FromEncoding(parameter, this.Info.ToPrivKey().ToBytes());
    signer.Init(true, privkey);
    signer.BlockUpdate(msg, 0, msg.Length);

    return (signer.GenerateSignature(), this.PubKey);
  }

  public static bool Verify(
    ReadOnlySpan<byte> pubkeyalice,
    ReadOnlySpan<byte> signature,
    ReadOnlySpan<byte> message,
    MLDsaParameters parameter)
  {
    var msg = message.ToArray();

    //Verify Signature
    var signer = new MLDsaSigner(parameter, true);
    var pubkey = MLDsaPublicKeyParameters.FromEncoding(
      parameter, pubkeyalice.ToArray());
    signer.Init(false, pubkey);
    signer.BlockUpdate(msg, 0, msg.Length);
    return signer.VerifySignature(signature.ToArray());
  }

  private static MLDsaParameters ToRngParameter()
  {
    var rand = new SecureRandom();
    var parameters = BcPqcServices.ToMLDsaParameters();
    var idx = rand.Next(parameters.Length);
    return parameters[idx];
  }

  public static MLDsaPublicKeyParameters ToPubKey(
    byte[] pubkey, MLDsaParameters parameter) =>
      MLDsaPublicKeyParameters.FromEncoding(parameter, pubkey);


  protected virtual void Dispose(bool disposing)
  {
    if (!this.IsDisposed)
    {
      if (disposing)
      {
      }
      this.IsDisposed = true;
    }
  }

  ~BobMLDSA() => Dispose(false);

  public void Dispose()
  {
    Dispose(true);
    GC.SuppressFinalize(this);
  }
}

File: PQC/BouncyCastle/MyBcPQC/MyBcPQC/TestMyBcPQCVb/TestBcPqcSignatur/Stateless/Test-ML-DSA/BobMLDSA.vb
Option Strict On
Option Explicit On



Imports michele.natale.BcPqcs
Imports michele.natale.Services
Imports Org.BouncyCastle.Security
Imports Org.BouncyCastle.Crypto.Signers
Imports Org.BouncyCastle.Crypto.Generators
Imports Org.BouncyCastle.Crypto.Parameters


Namespace michele.natale.TestBcPqcs

  Public Class BobMLDSA
    Implements IDisposable

    Private MPubKey As Byte() = Array.Empty(Of Byte)
    Public Property PubKey() As Byte()
      Get
        Return Me.MPubKey
      End Get
      Private Set
        Me.MPubKey = Value
      End Set
    End Property

    Private MIsDisposed As Boolean = True
    Public Property IsDisposed() As Boolean
      Get
        Return Me.MIsDisposed
      End Get
      Set
        Me.MIsDisposed = Value
      End Set
    End Property

    Private MInfo As MlDsaKeyPairInfo = Nothing
    Public Property Info() As MlDsaKeyPairInfo
      Get
        Return Me.MInfo
      End Get
      Private Set
        Me.MInfo = Value
      End Set
    End Property

    Private MParameter As MLDsaParameters = Nothing
    Public Property Parameter() As MLDsaParameters
      Get
        Return Me.MParameter
      End Get
      Private Set
        Me.MParameter = Value
      End Set
    End Property


    Private Rand As New SecureRandom

    Public Sub New()
      Me.New(ToRngParameter())
    End Sub

    Public Sub New(parameter As MLDsaParameters)
      Me.Parameter = parameter

      Dim generator = New MLDsaKeyGenerationParameters(Me.Rand, parameter)

      Dim keypair_generator = New MLDsaKeyPairGenerator()
      keypair_generator.Init(generator)

      Dim keypair = keypair_generator.GenerateKeyPair()
      Me.PubKey = DirectCast(keypair.[Public], MLDsaPublicKeyParameters).GetEncoded()
      Dim privkey = DirectCast(keypair.[Private], MLDsaPrivateKeyParameters).GetEncoded()
      Me.Info = New MlDsaKeyPairInfo(Me.PubKey, privkey, parameter)

      Array.Clear(privkey)
      Me.IsDisposed = False
    End Sub

    Public Sub Clear()

      If Me.MIsDisposed Then Return

      Me.Info.Dispose()
      If Me.PubKey IsNot Nothing Then
        Array.Clear(Me.PubKey)
      End If

      Me.Rand = Nothing
      Me.Info = Nothing
      Me.PubKey = Nothing
      Me.Parameter = MLDsaParameters.ml_dsa_44
    End Sub

    Public Function ToPubKey() As MLDsaPublicKeyParameters
      Return MLDsaPublicKeyParameters.FromEncoding(Me.Parameter, Me.PubKey)
    End Function

    Public Function Sign(message As Byte()) As (Sign As Byte(), Pubkey As Byte())
      'Sign Message-Data 

      'ML-DSA Parameter 
      Dim parameter = Me.Info.ToParameter()

      Dim signer = New MLDsaSigner(parameter, True)
      Dim privkey = MLDsaPrivateKeyParameters.FromEncoding(parameter, Me.Info.ToPrivKey().ToBytes())
      signer.Init(True, privkey)
      signer.BlockUpdate(message, 0, message.Length)

      Return (signer.GenerateSignature(), Me.PubKey)
    End Function

    Public Shared Function Verify(
    pubkeyalice As Byte(),
    signature As Byte(),
    message As Byte(),
    Parameter As MLDsaParameters) As Boolean

      'Verify Signature
      Dim signer = New MLDsaSigner(Parameter, True)
      Dim PubKey = MLDsaPublicKeyParameters.FromEncoding(
        Parameter, pubkeyalice)
      signer.Init(False, PubKey)
      signer.BlockUpdate(message, 0, message.Length)
      Return signer.VerifySignature(signature)
    End Function

    Private Shared Function ToRngParameter() As MLDsaParameters
      Dim rand = New SecureRandom()
      Dim parameters = BcPqcServices.ToMLDsaParameters()
      Dim idx = rand.[Next](parameters.Length)
      Return parameters(idx)
    End Function

    Public Shared Function ToPubKey(
      pubkey As Byte(), parameter As MLDsaParameters) As MLDsaPublicKeyParameters
      Return MLDsaPublicKeyParameters.FromEncoding(parameter, pubkey)
    End Function

    Protected Overridable Sub Dispose(disposing As Boolean)
      If Not Me.IsDisposed Then
        If disposing Then Me.Clear()
        Me.IsDisposed = True
      End If
    End Sub

    Protected Overrides Sub Finalize()
      Me.Dispose(False)
    End Sub

    Public Sub Dispose() Implements IDisposable.Dispose
      Me.Dispose(True)
      GC.SuppressFinalize(Me)
    End Sub
  End Class
End Namespace
